thunder/core/symbol.py

```diff
@@ -268,6 +268,21 @@ def __call__(self, *args, **kwargs):
             result = self.meta(*args, **kwargs)
             trace.pop_scope()
 
+        if self.tags is not None:
+            from thunder.core.prims import OpTags
+            from thunder.core.proxies import TensorProxy
+            from thunder.core.utils import check, check_type
+
+            if OpTags.SHAPE_OP in self.tags:
+                flat_result, _ = tree_flatten(result)
+                check(
+                    len(flat_result) == 1,
+                    lambda: f"{self.id=} has {self.tags} but {len(flat_result)} results, not 1",
+                )
+                result_t: TensorProxy = flat_result[0]
+                check_type(result_t, TensorProxy)
+                result_t.set_view_of(tree_flatten((args, kwargs))[0][0])
```

Comment:
This isn't correct. The result of a shape operation is not always a view of the original tensor

Reply:
you're right but it'd require strides information to tell if the output is a view or a copy

Reply:
> The result of a shape operation is not always a view of the original tensor

This is true only for three operations `reshape`, `reshape_as` and `flatten`.
PyTorch documentation says that users shouldn't rely on the result of `reshape` being a view or copy, the same is translated to `flatten` because `reshape` is used to implement it. In Thunder, `reshape` should be assumed to always produce a copy, if `reshape+in_place_op` is traced from the user code there can be a warning under a debug flag but not in the normal run.

There's also `contiguous` that might return the input itself. It might be better to error out for `contiguous+in_place_op` at first for the user to resolve the ambiguity.

Reply:
Refusing to execute programs with ambiguous inplace semantics seems totally fine, but we should consider that approach in a design review. There may be more operations than contiguous that detect when a no-op occurs and just return their input, too, so we should think what the semantics of those operations are.

Reply:
> There may be more operations than contiguous that detect when a no-op occurs and just return their input, too, so we should think what the semantics of those operations are.

What are these operations?

Reply:
`to` variations, we could have `a + 0` return `a`, `abs` on dtypes without negative values, `isfinite` on exact dtypes, ...

Reply:
> `to` variations, we could have `a + 0` return `a`, `abs` on dtypes without negative values, `isfinite` on exact dtypes, ...

These examples are resolvable at Thunder tracing time (except for zero addition) and eliminated by DCE today.

Reply:
Sure, but we should be clear that our design, then, is that if an operation decides to no-op it does not create a new storage (even if it otherwise would), and inplace operations on the result of the operation will affect the input's storage, too